external/fv3fit/fv3fit/reservoir/train.py

```diff
@@ -0,0 +1,213 @@
+
+
+@register_training_function("pure-reservoir", ReservoirTrainingConfig)
+def train_reservoir_model(
+    hyperparameters: ReservoirTrainingConfig,
+    train_batches: tf.data.Dataset,
+    validation_batches: Optional[tf.data.Dataset],
+) -> Predictor:
+
+    if hyperparameters.autoencoder_path is not None:
+        autoencoder = Autoencoder.load(hyperparameters.autoencoder_path)
+    else:
+        autoencoder = None  # type: ignore
+
+    # Standard scale data, calculating mean and std along the stacked x,y,z,feature dim
+    norm_batch = next(iter(train_batches))
+    norm_batch_concat = concat_variables_along_feature_dim(
+        variables=hyperparameters.input_variables, variable_tensors=norm_batch
+    )
+
+    sample_batch = (
+        _encode_columns(norm_batch_concat, autoencoder.encoder)
+        if autoencoder is not None
+        else norm_batch_concat
+    )
+    subdomain_config = hyperparameters.subdomain
+    rank_divider = RankDivider(
+        subdomain_layout=subdomain_config.layout,
+        rank_dims=subdomain_config.rank_dims,
+        rank_extent=tuple(sample_batch.shape),
+        overlap=subdomain_config.overlap,
+    )
+
+    # first data dim is time, the rest of the elements of the
+    # subdomain+halo are are flattened into feature dimension
+    n_input_features = int(
+        np.prod(rank_divider.get_subdomain_extent(with_overlap=True)[1:])
+    )
+    reservoir = Reservoir(
+        hyperparameters=hyperparameters.reservoir_hyperparameters,
+        input_size=n_input_features,
+    )
+
+    # One readout is trained per subdomain when iterating over batches,
+    # and they are combined after training into a CombinedReadout
+    subdomain_regressors = [
+        BatchLinearRegressor(hyperparameters.readout_hyperparameters)
+        for r in range(rank_divider.n_subdomains)
+    ]
+    for b, batch_data in enumerate(train_batches):
+        # reservoir increment occurs in this call, so always call this
+        # function even if X, Y are not used for readout training.
+        time_series_with_overlap, time_series_without_overlap = _process_batch_data(
+            variables=hyperparameters.input_variables,
+            batch_data=batch_data,
+            rank_divider=rank_divider,
+            autoencoder=autoencoder,
+        )
+        if b < hyperparameters.n_batches_burn:
+            logger.info(f"Synchronizing on batch {b+1}")
+        reservoir_state_time_series = _get_reservoir_state_time_series(
+            time_series_with_overlap, hyperparameters.input_noise, reservoir
+        )
+        # X has dimensions [time, reservoir_state, subdomain]
+        X_batch = reservoir_state_time_series[:-1]
+        if hyperparameters.square_half_hidden_state is True:
+            X_batch = square_even_terms(X_batch, axis=1)
+
+        # Y has dimensions [time, feature, subdomain] where feature dimension
+        # has flattened (x, y, z, variable) coordinates
+        Y_batch = time_series_without_overlap[1:]
```

Comment:
This code (starting from the first definition of `X_batch`) can all go under the `if b >= hyperparameters.n_batches_burn:`, correct?  It could be worth breaking off into a separate function.